### ISSUE TITLE
multi-core proof time & memory measurements

### ISSUE THREAD

**Issue by nathan-at-least** [zcash/zcash#1940]
Let's augment speed.z.cash to do performance tests for multiple cores on proving.

The simple first step (for this ticket) is to do a test with the number of cores present on the reference machine.

**Future Work:** (these should be separate tickets later; don't block closing this ticket on the following.)

Eventually it'd be nice to have one measurement for ``N`` cores from 1 to the machine limit on the reference machine.

In the future we could get a second very-high-core-count machine to serve as a second reference for this kind of metric. That kind of machine could make sense for cases with high proof-throughput needs.

**Comment by tromer** [comment]
Let's distinguish two user needs, and two tests:
* For "low proof-latency needs", you would use all cores in parallel for a single JoinSplit at a time.
* For "high proof-bandwidth needs", you can do the above (to minimize the memory working set), or you can run many single-threaded JoinSplits in parallel (to minimize cross-core communication and avoid some tradeoffs in the parallel FFT). Not obvious which will be faster.

**Comment by str4d** [comment]
The current codebase does not set `OMP_NUM_THREADS` (or call `omp_set_num_threads()`), meaning that [the existing benchmark](https://speed.z.cash/timeline/?exe=1&base=1%2B9&ben=time+createjoinsplit&env=1&revs=50&equid=off&quarts=on&extr=on) defaults to using as many internal threads as there are cores on the reference machine. We can trivially add two additional benchmarks:

- One that sets `OMP_NUM_THREADS=1` before the current benchmark.
- One that sets `OMP_NUM_THREADS=1` and then starts `NUM_CORES` JoinSplit threads.

**Comment by tromer** [comment]
Not the distinction between the number of physical cores, and the number of virtual CPUs detected by Linux, which is double the number of cores on Intel chips with HyperThreading enabled.
OpenMP defaults to the latter.

That's indeed the right choice for minimum-latency libsnark proving. (Some parallel code works faster with just one thread per core, using the whole L1 cache, but this isn't the case for libsnark).

I'm not sure what's the best choice for the case of many single-threaded libsnark provers.

**Comment by str4d** [comment]
In terms of benchmarks, do we want "seconds per JoinSplit"? I think that makes the most sense for comparison. Thus:

- Current benchmark is "seconds per JoinSplit" using four threads on one JoinSplit.
- First option above is "seconds per JoinSplit" using one thread on one JoinSplit.
- Second option above would be "seconds per JoinSplit" using four threads on four JoinSplits, so would need to average the four results and then divide by four to get the average "seconds per JoinSplit" for the *machine*.

**Comment by tromer** [comment]
Related to #1966, "Reduce memory consumption due to proving key".

**Comment by tromer** [comment]
Regarding parallel multiple JoinSplits vs. parallelized single JoinSplit, see #1282.
